airflow/utils/log/file_task_handler.py

```diff
@@ -0,0 +1,186 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+
+from airflow import configuration as conf
+from airflow.configuration import AirflowConfigException
+from airflow.utils.file import mkdirs
+
+class FileTaskHandler(logging.Handler):
+    """
+    FileTaskHandler is a python log handler that handles and reads
+    task instance logs. It creates and delegates log handling
+    to `logging.FileHandler` after receiving task instance context.
+    It reads logs from task instance's host machine.
+    """
+
+    def __init__(self, base_log_folder):
+        super(FileTaskHandler, self).__init__()
+        self.handler = None
+        self.local_base = os.path.expanduser(base_log_folder)
+
+    def set_context(self, task_instance):
+        """
+        Provide task_instance context to airflow task handler.
+        :param task_instance: task instance object
+        """
+        local_loc = self._init_file(task_instance)
+        self.handler = logging.FileHandler(local_loc)
+        self.handler.setFormatter(self.formatter)
+        self.handler.setLevel(self.level)
+
+    def emit(self, record):
+        if self.handler is not None:
+            self.handler.emit(record)
+
+    def flush(self):
+        if self.handler is not None:
+            self.handler.flush()
+
+    def close(self):
+        if self.handler is not None:
+            self.handler.close()
+
+    def _read(self, task_instance, try_number):
+        """
+        Template method that contains custom logic of reading
+        logs given the try_number.
+        :param task_instance: task instance record
+        :param try_number: current try_number to read log from
+        :return: log message as a string
+        """
+
+        # Task instance here might be different from task instance when
+        # initializing the handler. Thus explicitly getting log location
+        # is needed to get correct log path.
+        log_relative_path = self._get_log_relative_path(task_instance, try_number)
+        loc = os.path.join(self.local_base, log_relative_path)
+        log = ""
+
+        if os.path.exists(loc):
+            try:
+                with open(loc) as f:
+                    log += "*** Reading local log.\n" + "".join(f.readlines())
+            except Exception as e:
+                log = "*** Failed to load local log file: {}. {}\n".format(loc, str(e))
+        else:
+            url = os.path.join("http://{ti.hostname}:{worker_log_server_port}/log",
```

Comment:
what exactly does this `worker_log_server_port` do? If we remove it here would different workers be able to access each other's local disk to fetch logs?